R/html-tweak.R

```diff
@@ -129,6 +129,103 @@ tweak_footnotes <- function(html) {
   xml2::xml_remove(container)
 }
 
+tablist_item <- function(tab, html, parent_id) {
+  id <- xml2::xml_attr(tab, "id")
+  text <- xml_text1(xml2::xml_child(tab))
+  ul_nav <- xml2::xml_find_first(html, sprintf("//ul[@id='%s']", parent_id))
+
+  xml2::xml_add_child(
+    ul_nav,
+    "a",
+    text,
+    `data-toggle` = "tab",
+    href = paste0("#", id),
+    role = "tab"
+  )
+
+  if (grepl("active", xml2::xml_attr(tab, "class"))) {
+    class <- "active"
+  } else {
+    class = ""
+  }
+  # a's need to be wrapped in li's
+  xml2::xml_add_parent(
+    xml2::xml_find_first(html, sprintf("//a[@href='%s']", paste0("#", id))),
+    "li",
+    class = class
+  )
+}
+
+# For filling the content div of a tabset
+tablist_content <- function(tab, html, parent_id) {
+  # remove 1st child that is the header
+  xml2::xml_remove(xml2::xml_child(tab))
+
+  if (grepl("active", xml2::xml_attr(tab, "class"))) {
+    xml2::xml_attr(tab, "class") <- sprintf("tab-pane active")
+  } else {
+    xml2::xml_attr(tab, "class") <- sprintf("tab-pane")
+  }
+
+  xml2::xml_attr(tab, "role") <- "tabpanel"
+
+  content_div <- xml2::xml_find_first(
+    html,
+    sprintf("//div[@id='%s']/div", parent_id)
+  )
+
+  xml2::xml_add_child(content_div, tab)
+}
+
+tweak_tabset <- function(html) {
+  id <- xml2::xml_attr(html, "id")
+
+  # Users can choose pills or tabs
+  nav_class <- if (grepl("tabset-pills", xml2::xml_attr(html, "class"))) {
+    "nav-pills"
+  } else {
+    "nav-tabs"
+  }
+
+  # Add empty ul for nav
+  xml2::xml_add_child(html, "ul", class=sprintf("nav %s nav-row", nav_class), id=id)
+  # Add empty div for content
+  xml2::xml_add_child(html, "div", class="tab-content")
+
+  # Identify tabs and get them in an object
+  tabs <- xml2::xml_find_all(html, "div[contains(@id, 'tab')]")
+
+  # Remove tabs from original HTML
+  xml2::xml_remove(tabs)
+
+  # Fill the ul for nav
+  purrr::walk(tabs, tablist_item, html = html, parent_id = id)
+
+  # Fill the div for content
+  purrr::walk(tabs, tablist_content, html = html, parent_id = id)
+
+  # activate first tab unless another one is already activated
+  # (by {.active} in the source Rmd)
+  nav_url <- xml2::xml_find_first(html, sprintf("//ul[@id='%s']", id))
+  if (!any(grepl("active", xml2::xml_attr(xml2::xml_children(nav_url), "class")))) {
+    tweak_class_prepend(xml2::xml_child(nav_url), "active")
+  }
+
+  content_div <- xml2::xml_find_first(html, sprintf("//div[@id='%s']/div", id))
+  if (!any(grepl("active", xml2::xml_attr(xml2::xml_children(content_div), "class")))) {
+    tweak_class_prepend(xml2::xml_child(content_div), "active")
+  }
+}
+
+tweak_tabsets <- function(html) {
```

Comment:
Can you please arrange these functions in opposite order so the high level functions come first?